timelock-api/src/main/java/com/palantir/atlasdb/timelock/watch/LockWatchService.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.watch;
+
+import java.util.UUID;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+/**
+ * A service that allows users of a {@link com.palantir.lock.v2.TimelockService} to track what is happening with their
+ * locks. Specifically, users can register a watch over some lock IDs. Whenever a watched lock is locked or unlocked,
+ * an event is published to this service.
+ */
+@Path("/lock-watch")
+public interface LockWatchService {
```

Comment:
To summarize my rough understanding of usage:

- I must register a watch before I can be alerted of any changes, so should generally register a watch before populating my cache. If I must load my cache first to know what the keys are, then I should reread the values after setting the watches.
- To use the cache in a transaction, I must fetch my read timestamp before checking my watch states. In a serializable transaction, I must fetch my commit timestamp before checking watch states for read-write conflicts. 
- My `LockWatchState` response is the same size scale as my set of lock descriptors.

The last point is interesting because we might instead have considered a response with less granularity, at the expense of less granularity about which cache entries are stale. I'm not sure of any use cases off hand for batched lock watches, though, which is the only place this would particularly matter. 

Could we still avoid returning the lock descriptor in the response, given that the client presumably already knows the descriptor -> UUID mapping? I guess that only works when each watch is for a single lock/key.

Reply:
Correct: I only know that the value `V` I'm reading for some cell `C` hasn't changed if I knew that the watch for `C` was in some state `S` before I read `V`, and that the watch for `C` is still in state `S`. Thus point 1 is correct.

The intention was for this to replace a KVS read that a user might do in their transaction task, and yes the start/read timestamp _must_ have been retrieved by this point. Otherwise, we could read the cache, think our value is still current, and then have another transaction lock and commit before we get our read timestamp. This logic similarly holds for commit timestamps in serializable transactions, as you mention correctly.

The third point is correct, and yep there is some freedom with regard to granularity. I guess the idea was that if you were reading a cache that had say three values and one of them changed, you could register one watch, and then only need to re-read the one that changed. That said, this might be better achieved by allowing some form of a batched endpoint for retrieving lock state (rather than passing the descriptors around) - happy to change this if we decide we prefer that.